### PR TITLE
Jenkins 4.0.0 Release.

### PR DESCRIPTION
**Pull request by kaiwalyajoshi**
There are many significant updates in this release.
- Use of [USI](https://github.com/mesosphere/usi) as the scheduler for Mesos-Plugin.
- Support for Windows agents.
- Removal of `config.xml` in favour of JCasC.
- Ability to specify a user defined list of plugins to install.
- Base-Tech update to Jenkins 2.204.2 LTS
- Many cleanups for easier maintainabilty and stability.

### CODE REVIEW

File: conf/jenkins/configuration.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        agentUser: "${JENKINS_AGENT_USER:-root}"
+        frameworkName: "${JENKINS_FRAMEWORK_NAME:-Jenkins Scheduler}"
+        jenkinsURL: "http://${JENKINS_FRAMEWORK_NAME:-jenkins}.${MARATHON_NAME:-marathon}.mesos:${PORT0:-8080}"
+        mesosAgentSpecTemplates:
+          - label: "${JENKINS_LINUX_AGENT_LABEL:-linux}"
+            agentAttributes: "${JENKINS_LINUX_AGENT_OFFER_SELECTION_ATTRIBUTES:-}"
+            agentCommandStyle: Linux
+            containerInfo:
+              dockerForcePullImage: false
+              dockerImage: "${JENKINS_LINUX_AGENT_IMAGE:-mesosphere/jenkins-dind:0.6.0-alpine-experimental}"
```
Comment (jeschkies) [inline conf/jenkins/configuration.yaml]: Out of curiosity, why  are we using a dind image as default and not a simple official JDK image?

File: conf/jenkins/configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            maxExecutors: ${JENKINS_LINUX_AGENT_MAX_EXECUTORS:-1}
+            mem: ${JENKINS_LINUX_AGENT_MEM:-512}
+            minExecutors: ${JENKINS_LINUX_AGENT_MIN_EXECUTORS:-1}
+            mode: EXCLUSIVE
+          - label: "${JENKINS_WINDOWS_AGENT_LABEL:-windows}"
+            agentAttributes: "${JENKINS_WINDOWS_AGENT_OFFER_SELECTION_ATTRIBUTES:-os:windows}"
+            agentCommandStyle: Windows
+            containerInfo:
+              dockerForcePullImage: false
+              dockerImage: "${JENKINS_WINDOWS_AGENT_IMAGE:-mesosphere/jenkins-windows-node:latest}"
```
Comment (jeschkies) [inline conf/jenkins/configuration.yaml]: I think we should check in the [Dockerfile](https://github.com/mesosphere/jenkins-mesos-plugin/blob/master/nodes/Dockerfile.windows) for the Windows image.

File: scripts/run.sh [hunk 3]
```diff
@@ -1,13 +1,28 @@
...
+    echo "Installing additional plugins..."
+    export REF=/var/jenkins_home
+    export JENKINS_WAR=${JENKINS_FOLDER}/jenkins.war
+    CURL_OPTIONS="-fkL" /usr/local/bin/install-plugins.sh $JENKINS_OPT_ADDITIONAL_PLUGINS 2>&1
+    echo "Completed installing additional plugins..."
+else
+    echo "No additional plugins requested for installation..."
+fi
+
 export LD_LIBRARY_PATH=/libmesos-bundle/lib:/libmesos-bundle/lib/mesos:$LD_LIBRARY_PATH
```
Comment (jeschkies) [inline scripts/run.sh]: This should not be required anymore. The new plugin based on USI does not use `libmesos`.

File: scripts/run.sh [hunk 4]
```diff
@@ -1,13 +1,28 @@
...
-export JENKINS_SLAVE_AGENT_PORT=$PORT_AGENT
 export MESOS_NATIVE_JAVA_LIBRARY=$(ls /libmesos-bundle/lib/libmesos-*.so)
```
Comment (jeschkies) [inline scripts/run.sh]: This can go.

File: universe/marathon.json.mustache [hunk 5]
```diff
@@ -6,52 +6,77 @@
...
   "env": {
-      "JENKINS_AGENT_ROLE": "{{roles.jenkins-agent-role}}",
-      "JENKINS_AGENT_USER": "{{advanced.jenkins-agent-user}}",
+      "JENKINS_AGENT_ROLE": "{{service.roles.jenkins-agent-role}}",
+      "JENKINS_AGENT_USER": "{{jenkins-agent.jenkins-agent-user}}",
       "JENKINS_FRAMEWORK_NAME": "{{service.name}}",
       "MARATHON_NAME": "{{service.marathon-name}}",
-      {{#security.secret-name}}
+      {{#service.security.secret-name}}
         "DCOS_SERVICE_ACCOUNT_CREDENTIAL": { "secret": "serviceCredential" },
```
Comment (jeschkies) [inline universe/marathon.json.mustache]: This should be `DCOS_SERVICE_ACCOUNT_PRIVATE_KEY`. And the `DCOS_SERVICE_ACCOUNT` option is missing as well. I'm not sure we still need `MESOS_AUTHENTICATEE` and `MESOS_HTTP_AUTHENTICATEE`.

**Review by jeschkies — CHANGES_REQUESTED** [review]
This is pretty awesome. I have a few comments on the configuration.

File: conf/jenkins/configuration.yaml (hunk 1 above)
Comment (kaiwalyajoshi) [reply conf/jenkins/configuration.yaml]: The common use case for customers is to build docker images on the jenkins agents, the simple JDK image doesn't do this and will break the default workflow for many users.

File: conf/jenkins/configuration.yaml (hunk 2 above)
Comment (kaiwalyajoshi) [reply conf/jenkins/configuration.yaml]: We kept the images seperate from this repo so far: https://github.com/mesosphere/dcos-jenkins-dind-agent

Should we do the same for the Windows image if it gets complicated down the line?

Comment (jeschkies) [reply conf/jenkins/configuration.yaml]: What is the benefit of this? Any new engineer will have to dig through our repo to find it.

**Review by jeschkies — APPROVED** [review]
Nice 🎉 

Could we import the Dockerfiles for the Dind images and deprecated the other repo?

**Comment by kaiwalyajoshi** [comment]
I've added both the linux and windows Dockerfiles to build the agent images.
I'll cut a beta release from this branch and soak that, we'll merge any fixes we find from soaking.

**Review by DominikDary — APPROVED** [review]

**Comment by harryge00** [comment]
Exited to see this PR 🚀  when will Jenkins 2.204.2 LTS be available in Universe?

**Pull request merged**